src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportFileManager.cs

```diff
@@ -0,0 +1,134 @@
+                // it means we are "restarting" an export job. We have to make sure that each file
+                // has been opened on the ExportDestinationClient.
+                await _exportDestinationClient.OpenFileAsync(latestFile.FileUri, cancellationToken);
+            }
+
+            _isInitialized = true;
+        }
+
+        public async Task WriteToFile(string resourceType, string partId, byte[] data, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(resourceType, nameof(resourceType));
+
+            if (!_isInitialized)
+            {
+                await Initialize(cancellationToken);
+            }
+
+            ExportFileInfo fileInfo = await GetFile(resourceType, cancellationToken);
+            await _exportDestinationClient.WriteFilePartAsync(fileInfo.FileUri, partId, data, cancellationToken);
+
+            fileInfo.IncrementCount(data.Length);
+
+            // We need to create a new file if the current file has exceeded a certain limit.
+            if (fileInfo.CommittedBytes >= _maxFileSizeInBytes)
+            {
+                await CreateNewFileAndUpdateMappings(resourceType, fileInfo.Sequence + 1, cancellationToken);
+            }
+        }
+
+        private async Task<ExportFileInfo> GetFile(string resourceType, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(resourceType, nameof(resourceType));
+
+            if (!_resourceTypeToFileInfoMapping.TryGetValue(resourceType, out ExportFileInfo exportFileInfo))
+            {
+                // If it is not present in the mapping, we have to create the first file for this resource type.
+                // Hence file sequence will always be 1 in this case.
+                exportFileInfo = await CreateNewFileAndUpdateMappings(resourceType, fileSequence: 1, cancellationToken);
+            }
+
+            return exportFileInfo;
+        }
+
+        private async Task<ExportFileInfo> CreateNewFileAndUpdateMappings(string resourceType, int fileSequence, CancellationToken cancellationToken)
+        {
+            string fileName = _exportJobRecord.ExportFormat + ".ndjson";
+
+            string dateTime = _exportJobRecord.QueuedTime.UtcDateTime.ToString("s")
+                    .Replace("-", string.Empty, StringComparison.OrdinalIgnoreCase)
+                    .Replace(":", string.Empty, StringComparison.OrdinalIgnoreCase);
+
+            fileName = fileName.Replace(ExportFormatTags.Timestamp, dateTime, StringComparison.OrdinalIgnoreCase);
+            fileName = fileName.Replace(ExportFormatTags.Id, _exportJobRecord.Id, StringComparison.OrdinalIgnoreCase);
+            fileName = fileName.Replace(ExportFormatTags.ResourceName, resourceType, StringComparison.OrdinalIgnoreCase);
+            fileName = fileName.Replace(ExportFormatTags.Sequence, fileSequence.ToString(), StringComparison.OrdinalIgnoreCase);
```

Comment:
Having the sequence as part of the format concerns me since a user could make a format that doesn't contain a sequence and that would break our ability to export large data sets. Should we make sequence something we automatically append to the end of a file name?

Reply:
That's a good point. I misunderstood the way export format is used. I have updated the code so that we add a sequence to the end of the file irrespective of what format we are using (since that is what I was aiming for).